src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/ExponentialRetryPolicy.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+
+namespace Azure.Base.Http.Pipeline
+{
+    internal class ExponentialRetryPolicy : RetryPolicy
+    {
+        private readonly int[] _retriableCodes;
+
+        private readonly Func<Exception, bool> _exceptionFilter;
+
+        private readonly int _maxRetries;
+
+        private readonly TimeSpan _delay;
+
+        private readonly TimeSpan _maxDelay;
+
+        private Random _random;
+
+        public ExponentialRetryPolicy(int[] retriableCodes, Func<Exception, bool> exceptionFilter, int maxRetries, TimeSpan delay, TimeSpan maxDelay)
```

Comment:
The guidelines state that all codes in 500-599 range should be retried. Should we have two arrays to add and remove codes?

Reply:
Maybe we should change int[] to `Func<HttpMessage, bool>` delegate?

Reply:
Personally, I like the delegate approach with a default fallback to a known good instance.  I like the flexibility of passing a higher order function and I think the consistency with filtering of exceptions would also be nice.

Reply:
I find it strange that we would have both a delegate and ShouldXxx methods. It seems like two ways of doing extensibility.

Reply:
ShouldXxx also does delay calculation. Let's keep the current surface for now and see how AzConfigRetry policy would look like. It has both custom error codes and custom logic (x-ms-retryafter)